### ISSUE TITLE
Empty MTrk chunk breaks MIDI Import

### ISSUE THREAD

**Issue by derselbst** [muse-sequencer/muse#458]
When importing a MIDI File, which has a MTrk chunk of zero size, some events in the follwing MTrk chunk might be ommitted, resulting in uncorrect delta-delays. See an example MIDI here: 

[https://drive.google.com/open?id=0B7HVOcdsHTi3QU8tS092ZlU4eVU](https://drive.google.com/open?id=0B7HVOcdsHTi3QU8tS092ZlU4eVU)

When importing the MIDI you should get the follwing result:
![musemidiwrong](https://cloud.githubusercontent.com/assets/8152480/12376855/b57a8ee4-bd08-11e5-958a-008ab6d884db.png)

Note the MThd: in order to import this file, you have to set number-of-trks to only 2, although the file contains obviously 3 MTrks.

Removing the empty MTrk chunk at offset 0x388 should cause correct import result:
![musemidicorrect](https://cloud.githubusercontent.com/assets/8152480/12376879/7650efb4-bd09-11e5-9a09-481a2d5ef360.png)

**Comment by derselbst** [comment]
One minor question by the way: Is it normal the way you nowadays display ctrl-events like mainvolume, pitch or pan in those track canvases (or what do you call them?) - is this a bug or feature?
Back in Muse 2.1 they were displayed like:
![musetrackcanvas](https://cloud.githubusercontent.com/assets/8152480/12376935/bd18b416-bd0b-11e5-9a54-3b9de5d875c3.png)

**Comment by terminator356** [comment]
On January 17, 2016 02:21:12 AM Tom M. wrote:
> One minor question by the way: Is it normal the way you nowadays display
> ctrl-events like mainvolume, pitch or pan in those track canvases (or what
> do you call them?) - is this a bug or feature? Back in Muse 2.1 they were
> displayed like:
> ![musetrackcanvas](https://cloud.githubusercontent.com/assets/8152480/123769
> 35/bd18b416-bd0b-11e5-9a54-3b9de5d875c3.png)
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/muse-sequencer/muse/issues/458#issuecomment-172309970

Hi, I've downloaded the file and will take a look.

Thanks.
Tim.

**Comment by terminator356** [comment]
On January 20, 2016 03:56:12 AM you wrote:
> On January 17, 2016 02:21:12 AM Tom M. wrote:
> > One minor question by the way: Is it normal the way you nowadays display
> > ctrl-events like mainvolume, pitch or pan in those track canvases (or what
> > do you call them?) - is this a bug or feature? Back in Muse 2.1 they were
> > displayed like:
> > ![musetrackcanvas](https://cloud.githubusercontent.com/assets/8152480/1237
> > 69 35/bd18b416-bd0b-11e5-9a54-3b9de5d875c3.png)
> > 
> > 
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/muse-sequencer/muse/issues/458#issuecomment-172309970
> 
> Hi, I've downloaded the file and will take a look.
> 
> Thanks.
> Tim.

Hi Tom.
Please try it now.

- Midi import: Ignore empty tracks with zero data. Fixed corruption if so.
- Midi import/export: Delete allocated memory.

*New: Midi export dialog now has 'Use running status', and the 
  instr/mode + port/device metas can now be turned off 
  for a 'cleaner' file.


By using these new Settings, I can re-export your imported midi file 
 and it looks close to the original.

I'll try to fix your non-drawing issue. It's kinda weird at first glance.
The controllers /will/ draw if I import the re-exported file.
Seems like only the original import won't draw.

Thanks.
Tim.

**Comment by derselbst** [comment]
Ok, the provided midi file now imports. However strictly speaking, as noted above, this is **no valid** midi file, since its header specifies only 2 MTrk chunks although there are 3 in the file. Correcting the midi header MusE still throws an unexpected EOF error. Excuse the pedantry.

But the other two features seem to work fine :)

**Comment by terminator356** [comment]
On January 22, 2016 09:05:15 AM Tom M. wrote:
> Ok, the provided midi file now imports. However strictly speaking, as noted
> above, this is **no valid** midi file, since its header specifies only 2
> MTrk chunks although there are 3 in the file. 

Exactly. Whatever made this file is broken.

> Correcting the midi header
> MusE still throws an unexpected EOF error. Excuse the pedantry.

How's that? Shouldn't happen. Can you please post this file?

> But the other two features seem to work fine :)

The exporter was also not exporting note-off velocities.
That was fixed too.

Tim.

**Comment by derselbst** [comment]
>How's that? Shouldn't happen. Can you please post this file?

[Here you go](https://drive.google.com/open?id=0B7HVOcdsHTi3YVZ0emFkNzlWaFE), same file but corrected number of tracks at offset 0x0B to 3 ;)

**Comment by derselbst** [comment]
how about PR #461?

**Comment by terminator356** [comment]
On January 22, 2016 11:08:13 AM Tom M. wrote:
> >How's that? Shouldn't happen. Can you please post this file?
> 
> [Here you
> go](https://drive.google.com/open?id=0B7HVOcdsHTi3YVZ0emFkNzlWaFE), same
> file but corrected number of tracks at offset 0x0B to 3 ;)
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/muse-sequencer/muse/issues/458#issuecomment-174014441

Ha! A bit too zealous in my fix.

Small revert.
Turns out the caller ignores empty lists anyway, so the fix was simpler.
(Likely, at a bare minimum tracks should contain 'track end' meta 0xff 0x2f.)

Tested OK with /both/ files now. No errors.
(I was surprised when I saw it work with both.)

Note: I cannot support loading of the third track in the broken file, 
 at this moment. It is possible by switching to a token discovery model 
 like our song files, but that's lots more work. For now we /rely/ on that 
 track count number ;-)
Nor will I attempt to support creating empty tracks, I'll leave it at that.

Thanks.
Tim.
